### ISSUE TITLE
Ban redundant IIFEs

### ISSUE THREAD

**Issue by graingert** [eslint/eslint#8843]
**Please describe what the rule should do:**

Often people converting legacy `script` style JS into modules keep redundant IIFEs. Write a rule to ban these.

A redundant IIFE is an IIFE where the result is not used, or the result is statically analysable.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style 
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something X
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// bad
((global) => {
    const foo = 3;
    ham();
    return 7;
})(window);

// good
const window = global;
const foo = 3;
ham();

// bad
$(((global) => {
    const foo = 3;
    ham();
    return 7;
})(window));

// 'good'

const window = global;
const foo = 3;
ham();
$(7);

// bad
const res = ((global) => {
    const foo = 3;
    ham();
    return 7;
})(window);

// 'good'
const window = global;
const foo = 3;
ham();
const res = 7;

// good
const res = (() => {
    switch (foo) {
        case 'ham': return 'spam';
        case 'eggs': return 'bacon';
        default: return 'sauce';
    }
})();

// bad
const res = (() => {
    switch (foo) {
        case 'ham': return 'sauce';
        case 'eggs': return 'sauce';
        default: return 'sauce';
    }
})();

// good
const res = 'sauce';

// bad
const res = (() => {
    switch (foo) {
        case 'ham': return 'sauce';
        case 'eggs': return 'sauce';
        default: throw new Error('invalid breakfast item');
    }
})();

// good
const res = 'sauce';
switch(foo) {
    case 'ham': break;
    case 'eggs': break;
    default: throw new Error('invalid breakfast item');
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Seems pretty handy, doesn't depend on exotic syntax.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. I'm a bit confused about what this rule would do -- why would it report an error for one of these but not the other?

```js

// good
const res = (() => {
    switch (foo) {
        case 'ham': return 'spam';
        case 'eggs': return 'bacon';
        default: return 'sauce';
    }
})();

// bad
const res = (() => {
    switch (foo) {
        case 'ham': return 'sauce';
        case 'eggs': return 'sauce';
        default: throw new Error('invalid breakfast item');
    }
})();
```

**Comment by graingert** [comment]
@not-an-aardvark because the good one can't be done without iife, nested ternary or do. And nobody likes nested ternary and do isn't out yet

**Comment by not-an-aardvark** [comment]
Wouldn't that also be the case for the bad one, though?

**Comment by graingert** [comment]
@not-an-aardvark

It's subtle but because the function only has one return value it doesn't need an iife

```js
const res = 'sauce';
switch(foo) {
    case 'ham': break;
    case 'eggs': break;
    default: throw new Error('invalid breakfast item');
}
```

**Comment by not-an-aardvark** [comment]
I see, thanks for explaining.

Hmm, personally I'm not convinced that this rule is worth adding to core -- it seems like a very user-specific preference. However, let's see what the rest of the team thinks.

**Comment by ilyavolodin** [comment]
I agree. It seems very situational and I can't say that I ran into a situation like that before. Feels like it might be a better fit for a plugin.

**Comment by not-an-aardvark** [comment]
It seems like the team is not in favor of this rule, so I'm going to close this issue. However, you can always implement the rule as part of a plugin.
